### PR TITLE
Heatmap enhancements

### LINKED ISSUE #430: Heatmap enhancements
- [ ] Draw border for highlighted rect on higher layer to achieve better performance (don't pass `isHovered` down and check in every `HeatmapRect` to determine rendering)
- [ ] Legend component for Heatmap (min-max values presented on colour scale with two values in between)
- [ ] Bug: after selecting Heatmap, LineChart Legend stays on screen
- [ ] Show dimensions name on Heatmap
- [ ] Clicking on dimension labels highlight should be proposed, clicking on rect, highlight should be proposed for both dimensions

### PR DESCRIPTION
**Pull request by adrianmroz**
Closes #430

Add highlight to Heatmap
Draw hover indicator on top of Heatmap and limit renders on HeatmapRectangles
Show Legend on Heatmap
Add zip function util

### CODE REVIEW

File: src/client/utils/dataset/dataset.mocha.ts [hunk 1]
```diff
@@ -18,43 +18,46 @@
...
 const timezone = Timezone.UTC;
 
+const mockSplit = (reference: string, type: SplitType, direction: SortDirection) =>
+  new Split({ reference, type, sort: new DimensionSort({ reference, direction }) });
+
 describe("Dataset", () => {
   it("works", () => {
-    expect(fillDatasetWithMissingValues(Dataset.fromJS(rawDataset.data[0][SPLIT] as Dataset), "pv_count", "site", orderByValueDecreasing, timezone).toJS())
+    const siteSplit = mockSplit("site", SplitType.string, SortDirection.descending);
+    expect(fillDatasetWithMissingValues(Dataset.fromJS(rawDataset.data[0][SPLIT] as Dataset), "pv_count", siteSplit, timezone).toJS())
```
Comment (erykpiast) [inline src/client/utils/dataset/dataset.mocha.ts]: I'd argue for moving what's inside `expect` clause to a named constant - it's going to be far more readable.

File: src/client/utils/dataset/dataset.mocha.ts [hunk 2]
```diff
@@ -18,43 +18,46 @@
...
+    const timeSplit = mockSplit("__time", SplitType.time, SortDirection.descending);
+    expect(fillDatasetWithMissingValues(Dataset.fromJS(rawDatasetWithTimeDimension.data[0][SPLIT] as Dataset), "click", timeSplit, timezone).toJS())
+      .to.deep.equal(Dataset.fromJS(reversedDatasetWithTimeDimension.data[0][SPLIT] as Dataset).toJS());
   });
 
   it("works with number ranges", () => {
-    expect(fillDatasetWithMissingValues(Dataset.fromJS(rawDataWithNumberRanges.data[0][SPLIT] as Dataset), "added", "deltaBucket100", orderByNumberRangeDimensionIncreasing, timezone).toJS())
-    .to.deep.equal(Dataset.fromJS(rawDataWithNumberRanges.data[0][SPLIT] as Dataset).toJS());
+    const deltaSplit = mockSplit("deltaBucket100", SplitType.number, SortDirection.ascending);
+    expect(fillDatasetWithMissingValues(Dataset.fromJS(rawDataWithNumberRanges.data[0][SPLIT] as Dataset), "added", deltaSplit, timezone).toJS())
```
Comment (erykpiast) [inline src/client/utils/dataset/dataset.mocha.ts]: It'd be valuable to hide that `Dataset.fromJS(_.data[0][SPLIT] as Dataset)` boilerplate behind some helper layer, but that's, as always, just my humble opinion :P

File: src/client/utils/dataset/dataset.ts [hunk 3]
```diff
@@ -17,12 +17,14 @@
 
 import { Timezone } from "chronoshift";
 import { Dataset, Datum, NumberRange, TimeRange } from "plywood";
+import { SortDirection } from "../../../common/models/sort/sort";
+import { Split, SplitType } from "../../../common/models/split/split";
 import { formatValue } from "../../../common/utils/formatter/formatter";
 import { SPLIT } from "../../config/constants";
 
-export type Order<D> = (a: [string, number, D], b: [string, number, D]) => number;
+type Order<D> = (a: [string, number, D], b: [string, number, D]) => number;
```
Comment (erykpiast) [inline src/client/utils/dataset/dataset.ts]: What kind of sorcery is this? Where the third element of the tuple goes in `orderByValueDecreasing` definition? Is it simply omitted and that's just fine for TS?

File: src/client/visualizations/base-visualization/base-visualization.tsx [hunk 4]
```diff
@@ -202,10 +203,10 @@ export class BaseVisualization<S extends BaseVisualizationState> extends React.C
     this.setState({ highlight: null });
   };
 
-  protected highlight(clauses: List<FilterClause>, key: string) {
+  protected highlight = (clauses: List<FilterClause>, key: string | null = null) => {
```
Comment (erykpiast) [inline src/client/visualizations/base-visualization/base-visualization.tsx]: There should be some shortcut syntax for this: `string | null = null`.

Comment (erykpiast) [reply src/client/visualizations/base-visualization/base-visualization.tsx]: Or maybe I'm confusing runtime values with types. Is `null` on the right side of `=` a value or a type?

File: src/client/visualizations/heat-map/heat-map.tsx [hunk 5]
```diff
@@ -20,162 +20,66 @@
...
 
-type HeatmapState = BaseVisualizationState & { preparedDataset: Dataset };
+interface HeatmapState extends BaseVisualizationState {
+  preparedDataset: Dataset;
+}
 
 export class HeatMap extends BaseVisualization<HeatmapState> {
   protected className = HEAT_MAP_MANIFEST.name;
 
+  getScales = memoizeOne(scales);
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heat-map.tsx]: Do you have some issues with React lifecycle or what? For what reason to memoize only the last call?

File: src/client/visualizations/heat-map/heatmap-corner.scss [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+.heatmap-corner-row-title {
+  position: absolute;
+  bottom: 5px;
+}
+
+.heatmap-corner-column-title {
+  position: absolute;
+  top: 0;
+  right: 12px;
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-corner.scss]: I'm a bit afraid to ask where these magic numbers come from... Are they related to the font-size or some other elements of the UI?

File: src/client/visualizations/heat-map/heatmap-corner.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  height: number;
+  colorScale: ColorScale;
+}
+
+export const HeatmapCorner: React.SFC<HeatmapCornerProps> = ({ colorScale, width, height, essence }) => {
+  const { dataCube, splits: { splits } } = essence;
+
+  const row = splits.get(0);
+  const column = splits.get(1);
+  const rowTitle = row.getTitle(dataCube.getDimension(row.reference));
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-corner.tsx]: If both `row` and `column` have some model in common, it looks like a method to me: `rowOrColumn.getTitleForDimension(dc: DataCube)`.

File: src/client/visualizations/heat-map/heatmap-highlight-modal.tsx [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  title: string;
+  stage: Stage;
+  layout: ScrollerLayout;
+  scroll: { left: number, top: number };
+  position: HighlightPosition;
+  dropHighlight: Fn;
+  acceptHighlight: Fn;
+}
+
+function calculateLeft(props: HeatmapHighlightModalProps): number {
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-highlight-modal.tsx]: Look ma, no tests!

File: src/client/visualizations/heat-map/heatmap-hover-tooltip.tsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+interface ContentProps {
+  datum: Datum;
+  showComparison: boolean;
+  series: ConcreteSeries;
+}
+
+const Content: React.SFC<ContentProps> = props => {
+  const { showComparison, series, datum } = props;
+  if (!showComparison) {
+    return <React.Fragment>{series.formatValue(datum)}</React.Fragment>;
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-hover-tooltip.tsx]: Why do you need a fragment here? Types?

File: src/client/visualizations/heat-map/heatmap-labels.tsx [hunk 10]
```diff
@@ -16,41 +16,52 @@
...
       return;
     }
 
-    const {
-      onMaxLabelSize = () => {}
-    } = this.props;
+    const { onMaxLabelSize = noop } = this.props;
 
-    const maxWidth = Array.from(this.container.querySelectorAll(".heatmap-label"))
+    const maxWidth = Array.from(this.container.current.querySelectorAll(".heatmap-label"))
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-labels.tsx]: Move `heatmap-label` to a constant, you have to synchronize this value with JSX.

File: src/client/visualizations/heat-map/heatmap-legend.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+export const HeatmapLegend: React.SFC<HeatmapLegendProps> = ({ height, series, scale }) => {
+  const [min, max] = scale.domain();
+  if (isNaN(min) || isNaN(max)) return null;
+
+  const [startColor, endColor] = scale.range();
+
+  const format = series.formatter();
+  const [y1, y2] = [10, height - 10];
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-legend.tsx]: Move `10` to some descriptively-named constant.

File: src/client/visualizations/heat-map/heatmap-legend.tsx [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        <stop offset="0%" stop-color={startColor} />
+        <stop offset="10%" stop-color={startColor} />
+        <stop offset="90%" stop-color={endColor} />
+        <stop offset="100%" stop-color={endColor} />
+      </linearGradient>
+    </defs>
+    <rect className="heatmap-legend-stripe" x={5} y={y1} width={10} height={y2 - y1} fill="url(#heatmap-stripe)" />
+    <line x1={5.5} x2={5.5} y1={y1} y2={y2} className="heatmap-legend-stripe-axis"/>
+    <g className="heatmap-lower-bound">
+      <line x1={5} x2={20} y1={y1 + 0.5} y2={y1 + 0.5} className="heatmap-lower-bound-tick"/>
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-legend.tsx]: cooonstants, please.

File: src/client/visualizations/heat-map/heatmap-rectangles.scss [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+  &::after {
+    position: absolute;
+    top: 0;
+    bottom: 0;
+    left: 0;
+    right: 0;
+    content: '';
+    pointer-events: none;
+    box-shadow: inset -45px -45px 30px -30px #fff;
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/heatmap-rectangles.scss]: some names would be nice.

File: src/client/visualizations/heat-map/labeled-heatmap.tsx [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+  resetHover = () => this.setState({ hoverPosition: null });
+
+  saveScroll = (scrollTop: number, scrollLeft: number) => this.setState({ scrollLeft, scrollTop });
+
+  saveLeftLabelWidth = (maxLabelWidth: number) => this.setState({ leftLabelsWidth: Math.min(maxLabelWidth, MAX_LEFT_LABELS_WIDTH) });
+
+  saveTopLabelHeight = (maxLabelHeight: number) => this.setState({ topLabelsHeight: Math.min(maxLabelHeight, MAX_TOP_LABELS_HEIGHT) });
+
+  handleHighlight = (x: number, y: number, part: ScrollerPart) => {
+    if (part === "body" || part === "top-gutter" || part === "left-gutter") {
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/labeled-heatmap.tsx]: If you revert the condition, there will be less code indented. Just saying.

File: src/client/visualizations/heat-map/labeled-heatmap.tsx [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+      const { saveHighlight, essence, dataset } = this.props;
+      const layout = this.layout();
+      const clauses = createHighlightClauses({ x: x - layout.left, y: y - layout.top, part }, essence, dataset);
+      saveHighlight(List(clauses));
+    }
+  };
+
+  private layout(): ScrollerLayout {
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/labeled-heatmap.tsx]: Some tests...?

File: src/client/visualizations/heat-map/utils/create-highlight-clauses.ts [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+function secondSplitSelection(leftOffset: number, essence: Essence, dataset: Datum[]): SplitSelection {
+  const { dataCube, splits: { splits } } = essence;
+  const dimensionName = splits.get(1).reference;
+  const dimension = dataCube.getDimension(dimensionName);
+  const labelIndex = Math.floor(leftOffset / TILE_SIZE);
+  const value = (dataset[0][SPLIT] as Dataset).data[labelIndex][dimensionName];
+  return { value, dimension };
+}
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/utils/create-highlight-clauses.ts]: Looks like a place to dry, but I don't have idea how. Some high-order function accepting split index and a function for extracting `value`?

File: src/client/visualizations/heat-map/utils/create-highlight-clauses.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+function splitSelectionToClause({ value, dimension: { kind, name: reference } }: SplitSelection): FilterClause {
+  switch (kind) {
+    case "string":
+      return new StringFilterClause({ reference, action: StringFilterAction.IN, values: Set.of(String(value)) });
+    case "boolean":
+      return new BooleanFilterClause({ reference, values: Set.of(value as Booleanish) });
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/utils/create-highlight-clauses.ts]: Booleanish ❤️

File: src/client/visualizations/heat-map/utils/equal-props.mocha.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+ * limitations under the License.
+ */
+
+import { expect } from "chai";
+import { HeatMapRectanglesProps } from "../heatmap-rectangles";
+import { equalProps } from "./equal-props";
+
+describe("equalProps", () => {
+  it("should return true if all props are the referentially identical", () => {
+    const props = { propString: "foo", propNumber: 42, propBoolean: true, propObject: { bazz: true } } as any as HeatMapRectanglesProps;
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/utils/equal-props.mocha.ts]: Some new lines, please...?

File: src/client/visualizations/heat-map/utils/equal-props.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+import { isEqualable } from "../../../../common/utils/immutable-utils/immutable-utils";
+import { HeatMapRectanglesProps } from "../heatmap-rectangles";
+
+function equals(a: any, b: unknown): boolean {
+  if (a === null) {
+    return a === b;
+  }
+  if (isEqualable(a)) {
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/utils/equal-props.ts]:
```javascript
if (a !== null && isEqualable(a)) { ... }
return a === b;
```

File: src/client/visualizations/heat-map/utils/equal-props.ts [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+    return a.equals(b);
+  }
+  return a === b;
+}
+
+export function equalProps(oldProps: HeatMapRectanglesProps, newProps: HeatMapRectanglesProps): boolean {
+  const keys = Object.keys(oldProps) as Array<keyof HeatMapRectanglesProps>;
+  return keys.reduce((result, key) => {
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/utils/equal-props.ts]: Looks like `.every` to me.

File: src/common/utils/functional/functional.mocha.ts [hunk 21]
```diff
@@ -44,6 +44,30 @@ describe("Functional utilities", () => {
     });
   });
 
+  describe.only("zip", () => {
```
Comment (erykpiast) [inline src/common/utils/functional/functional.mocha.ts]: 😱

**Comment by mkuthan** [comment]
Horizontal split label is misaligned.

http://tinyurl.com/rqdktbx
image

**Comment by mkuthan** [comment]
Horizontal highlight is out of vis area:

image

File: src/client/visualizations/base-visualization/base-visualization.tsx (hunk 4 above)
Comment (adrianmroz) [reply src/client/visualizations/base-visualization/base-visualization.tsx]: It’s value. Nullable string, by default it’s null

File: src/client/visualizations/heat-map/heat-map.tsx (hunk 5 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heat-map.tsx]: If this function params change, there’s a lot of work to do elsewhere, and savings there would have minimal impact. That way I don’t need to handle buffer of values.

File: src/client/visualizations/heat-map/heatmap-corner.scss (hunk 6 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heatmap-corner.scss]: This is result of function of font size. Non trivial one.

File: src/client/visualizations/heat-map/heatmap-hover-tooltip.tsx (hunk 9 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heatmap-hover-tooltip.tsx]: Hmm, dunno.

File: src/client/visualizations/heat-map/heatmap-legend.tsx (hunk 11 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heatmap-legend.tsx]: `ten`?

File: src/client/visualizations/heat-map/heatmap-rectangles.scss (hunk 13 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heatmap-rectangles.scss]: Like what, `insetTop`?

File: src/common/utils/functional/functional.mocha.ts (hunk 21 above)
Comment (adrianmroz) [reply src/common/utils/functional/functional.mocha.ts]: That’s why we have code review!

File: src/client/utils/dataset/dataset.ts (hunk 3 above)
Comment (adrianmroz) [reply src/client/utils/dataset/dataset.ts]: It is in type definition via `Order` type. In Implementation we don’t use it so we don’t destructure for it. It’s more like ... JavaScript :)

File: src/client/visualizations/heat-map/utils/equal-props.ts (hunk 19 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/utils/equal-props.ts]: This condition is harder to read. We have three cases: null, object with equal method, reference value. This three branches.

File: src/client/visualizations/heat-map/heatmap-corner.tsx (hunk 7 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heatmap-corner.tsx]: I don’t like either of those and will not bother changing :o

File: src/client/visualizations/heat-map/heatmap-highlight-modal.tsx (hunk 8 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heatmap-highlight-modal.tsx]: One case for each branch? It’s math so just checking if x equals specific number?

File: src/client/visualizations/heat-map/utils/create-highlight-clauses.ts (hunk 16 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/utils/create-highlight-clauses.ts]: More like composition. Outer function would pick split, offset and Dataset and pass rest.

File: src/client/visualizations/heat-map/labeled-heatmap.tsx (hunk 15 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/labeled-heatmap.tsx]: What cases? Three for each clamp? And one for rest of the props (no branching)?

File: src/client/visualizations/heat-map/heatmap-hover-tooltip.tsx (hunk 9 above)
Comment (adrianmroz) [reply src/client/visualizations/heat-map/heatmap-hover-tooltip.tsx]: Yup, React typings require that we have `ReactElement` here and string is not part of this type.

File: src/client/visualizations/heat-map/utils/highlight-modal-position.mocha.ts [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+    height: 200,
+    width: 300,
+    x: 20,
+    y: 50
+  })
+};
+
+const mockProps = (props: Partial<CoordinatesProps> = {}): CoordinatesProps => ({ ...defaults, ...props });
+
+describe.only("highlightModalPosition", () => {
```
Comment (erykpiast) [inline src/client/visualizations/heat-map/utils/highlight-modal-position.mocha.ts]: sci!

Comment (adrianmroz) [reply src/client/visualizations/heat-map/utils/highlight-modal-position.mocha.ts]: shame :bell: shame :bell:

**Pull request merged**